app/lib/api/api_util.dart

```diff
@@ -109,37 +112,130 @@ String getFacePreviewUrlRelative(
   return "index.php/apps/facerecognition/face/$faceId/thumb/$size";
 }
 
-/// Query the app password for [account]
-Future<String> exchangePassword(Account account) async {
-  final response = await Api(account).request(
-    "GET",
-    "ocs/v2.php/core/getapppassword",
+/// Initiate a login with Nextclouds login flow v2: https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html#login-flow-v2
+Future<InitiateLoginResponse> initiateLogin(Uri uri) async {
+  final response = await Api.fromBaseUrl(uri).request(
+    "POST",
+    "index.php/login/v2",
     header: {
-      "OCS-APIRequest": "true",
+      "User-Agent": "nc-photos",
     },
   );
   if (response.isGood) {
-    try {
-      final appPwdRegex = RegExp(r"<apppassword>(.*)</apppassword>");
-      final appPwdMatch = appPwdRegex.firstMatch(response.body);
-      return appPwdMatch!.group(1)!;
-    } catch (_) {
-      // this happens when the address is not the base URL and so Nextcloud
-      // returned the login page
-      throw InvalidBaseUrlException();
-    }
-  } else if (response.statusCode == 403) {
-    // If the client is authenticated with an app password a 403 will be
-    // returned
-    _log.info("[exchangePassword] Already an app password");
-    return account.password;
+    return InitiateLoginResponse.fromJsonString(response.body);
   } else {
     _log.severe(
-        "[exchangePassword] Failed while requesting app password: $response");
+        "[initiateLogin] Failed while requesting app password: $response");
     throw ApiException(
         response: response,
-        message: "Server responed with an error: HTTP ${response.statusCode}");
+        message: "Server responded with an error: HTTP ${response.statusCode}");
+  }
+}
+
+/// Retrieve App Password after successful initiation of login with Nextclouds login flow v2: https://docs.nextcloud.com/server/latest/developer_manual/client_apis/LoginFlow/index.html#login-flow-v2
+Future<AppPasswordResponse> _getAppPassword(
+    InitiateLoginPollOptions options) async {
+  Uri baseUrl;
+  if (options.endpoint.scheme == "http") {
+    baseUrl = Uri.http(options.endpoint.authority);
```

Comment:
It seems that the API was only changed in flutter 3.3 and now the second param is optional (it wasn't in 3.0 I'm using). I'll make the upgrade to 3.3 then.

Reply:
done